partiql-parser/src/peg.rs

```diff
@@ -4,12 +4,63 @@
 //! can be exported for users to consume.
 
 use crate::prelude::*;
-use pest::Parser;
+use crate::result::syntax_error;
+use pest::iterators::{Pair, Pairs};
+use pest::{Parser, RuleType};
 use pest_derive::Parser;
 
 #[derive(Parser)]
 #[grammar = "partiql.pest"]
-struct PartiQLParser;
+pub(crate) struct PartiQLParser;
+
+/// Extension methods for working with [`Pairs`].
+pub(crate) trait PairsExt<'val, R: RuleType> {
+    /// Consumes a [`Pairs`] as a singleton, returning an error if there are less or more than
+    /// one [`Pair`].
+    fn exactly_one(self) -> ParserResult<Pair<'val, R>>;
+}
+
+impl<'val, R: RuleType> PairsExt<'val, R> for Pairs<'val, R> {
+    fn exactly_one(mut self) -> ParserResult<Pair<'val, R>> {
+        match self.next() {
+            Some(pair) => {
+                // make sure there isn't something more...
+                if let Some(other_pair) = self.next() {
+                    syntax_error(
+                        format!("Expected one token pair, got: {:?}, {:?}", pair, other_pair),
+                        pair.start_loc().into(),
+                    )?;
+                }
+                Ok(pair)
+            }
+            None => syntax_error(
+                "Expected at one token pair, got nothing!",
+                Position::Unknown,
```

Comment:
FIXME?

Reply:
sorry that was a lame comment!

is it always the case there is no position? I guess it could be none because you're at EOF or something..?

Reply:
The only way this could happen is if I wrote a Pest grammar with a rule that matched something but it and all sub-rules were silent and did not produce a `Pair` at all.  This would mean that I as the rule writer wrote something pretty insane and I think it is almost certainly a bug.

Add this to the reason why I think Pest is a little too low-level for my tastes as far as its APIs are concerned (though I think logically consistent).